regression/contracts/invar_havoc_dynamic_multi-dim_array_all_const_idx/main.c

```diff
@@ -17,7 +17,7 @@ void main()
   data[4][5][6] = 0;
 
   for(unsigned i = 0; i < SIZE; i++)
-    __CPROVER_loop_invariant(i <= SIZE)
+    __CPROVER_assigns(i, data[4][5][6]) __CPROVER_loop_invariant(i <= SIZE)
```

Comment:
As of now, yes, users can either rely on automatic inference entirely or manual annotation entirely. If users manually provide an assigns clause, it has to capture (at least) everything that's assigned in the loop, just like function contracts.

In future, we could find a way to combine automatically inferred things with user annotations, both for function and loop contracts.

Reply:
For function contracts, user don't have to specify formal parameters and local variables in the assigns clauses. So, I was expecting that we wouldn't have to specify local variables for loop contracts. Don't we automatically add all regions from `DECL` instructions on the write set? Again, this is not a blocker, but I don't think it's related to the automatic inference procedure we use either.

Reply:
Local variables (variables declared inside the loop / function _body_) are completely unrelated here. They are always automatically added to the write set and are not specified in the assigns clause (because they are not even visible in the enclosing scope). This is completely unrelated to loop counter declaration though. It's not declared inside the loop body.

Formal parameters are _somewhat_ related. One could see a `for` loop with `i_1`, ..., `1_n` initialization declarations as a lambda function that has `i_1`, ..., `1_n` as formal parameters and is invoked with the respective initialization values.
This could be implemented. I can grab all declarations from the initialization clause of a `for` loop and add them to the assigns clause automatically.